Bugfixes and style optimization

tillschweneker opened the pull request:
Hey @SimeonGriggs 👋 

I have made some more changes to the plugin. 

### Update dependencies
All dependencies are up-to-date now. Only React is still on 17. 

### Use framer motion to get x/y values
Instead of parsing the transform of the hotspot, framer motions' useMotionValue hook is used to get the correct value

### Allow changing hotspots after initial save
Before it was not possible to update hotspots after they where initially created, because the referenced imageRef did not exists in the DOM when the Spot components where created. And as refs do not trigger a re-render, the Spot components never knew about the image.
Simply switching the order of the Spot components and the image would have done the trick, but with the next change I was able to get rid of the imageRef altogether.

### Use DOMRect instead of ImageRef for bounds
The rectangle is tracked by a ResizeObserver and updates the bounds, when the browser window and image resizes. This way the hotspots are always correctly repositioned. The rect is also used to calculate the initial position of the hotspot. 

### Update hotspot styles 

- The hotspot is now semi-opaque and has a dot in the middle, to pinpoint the hotspot position. 
- The cursor is removed while dragging, so it does not distract.
- The hotspot is moved half its width to the top and left, to accurately display the position.

Commit:
chore: update dependencies

Commit:
multiple fixes and refactorings

- use framer motion to get values
- allow changing hotspots after initial save
- update hotspot style for easier targeting
- use resizeObserver to update hotspot positions when image size changes

robinpyon commented:
Hey @tillschweneker, thanks for this PR!

There is a little bit of noise here – are you able add a commit with this output run through prettier? (Just so we can get a cleaner diff)

I did give this a spin and it looks like the main issues of not being able update hotspots (once the document has been saved) as well those resizing issues are fixed – great stuff!

I did pick up on one regression which I believe stems from [eagerly memoizing `imageHotspotPathRoot`](https://github.com/sanity-io/sanity-plugin-hotspot-array/pull/5/files#diff-a9eecd9ace4d29335f4c096c606611f3b2ca658820682d869c4ff8010be4fc37R42). The rendered image will no longer update correctly when the image itself is changed, removed or otherwise updated (e.g. if it's cropped in the studio).

Commit:
chore: autoformat

Commit:
feat: Update hotspot styles

- The hotspot is now semi-opaque and has a dot in the middle, to pinpoint the hotspot position.
- The cursor is removed while dragging, so it does not distract.
- The hotspot is moved half its width to the top and left, to accurately display the position.

Commit:
fix: set dragElastic to 0 to prevent dragging the hotspot out of bounds

Commit:
feat: use framer motion to get x/y values

Instead of parsing the transform of the hotspot, framer motions' useMotionValue hook is used to get the correct value

Commit:
chore: convert indentation to spaces

Commit:
reafactor: add some types

Commit:
change: use DOMRect instead of ImageRef for bounds

The rectangle is tracked by a ResizeObserver and updates the bounds, when the browser window and image resizes. This way the hotspots are always correctly repositioned. The rect is also used to calculate the initial position of the hotspot.

This change also fixes the bug, that hotspot couldn't be changed after initial save

Commit:
chore: typecast lodash get to string to fix linter warning

Commit:
chore: autoformat

Commit:
fix: remove check for missing or non-number x or y

- the checks prevented hotspots wher x or y had the position 0
- by using a rect for bounds and framer motion to get the position these checks are superflous

Commit:
chore: set type of newRow

Commit:
fix: update imageUrl when any changes to the image occur

Commit:
revert: prepare for better diff

Commit:
Merge branch 'diff'

robinpyon committed:
chore: prettier

robinpyon committed:
refactor: don't use inline arrow functions, only conditionally apply new styles on drag

robinpyon commented:
Thank you @tillschweneker!
